Mathlib/AlgebraicGeometry/Morphisms/Constructors.lean

```diff
@@ -56,136 +56,157 @@ instance AffineTargetMorphismProperty.diagonal_respectsIso (P : AffineTargetMorp
+  · infer_instance
+  · infer_instance
+#align algebraic_geometry.affine_target_morphism_property.diagonal_of_target_affine_locally AlgebraicGeometry.HasAffineProperty.affineProperty_diagonal_of_diagonal_of_isPullback
+
+lemma _root_.CategoryTheory.IsPullback.of_id_fst {C} [Category C] {X Y : C} (f : X ⟶ Y) :
+    IsPullback (𝟙 _) f f (𝟙 _) := IsPullback.of_horiz_isIso ⟨by simp⟩
+
+lemma _root_.CategoryTheory.IsPullback.of_id_snd {C} [Category C] {X Y : C} (f : X ⟶ Y) :
+    IsPullback f (𝟙 _) (𝟙 _) f := IsPullback.of_vert_isIso ⟨by simp⟩
+
+theorem HasAffineProperty.of_affineProperty_diagonal
+    (P) {Q} [HasAffineProperty P Q] {X Y} {f : X ⟶ Y} [IsAffine Y] :
+    Q.diagonal f ↔ P.diagonal f := by
+  letI := isLocal_affineProperty P
+  refine ⟨fun hf ↦ ?_, affineProperty_diagonal_of_diagonal_of_isPullback P (.of_id_fst _)⟩
+  rw [← Q.diagonal.cancel_left_of_respectsIso
+    (pullback.fst (f := f) (g := 𝟙 Y)), pullback.condition, Category.comp_id] at hf
+  let 𝒰 := X.affineCover.pushforwardIso (inv (pullback.fst (f := f) (g := 𝟙 Y)))
+  have (i) : IsAffine (𝒰.obj i) := by dsimp [𝒰]; infer_instance
+  exact HasAffineProperty.diagonal_of_openCover P f (Scheme.openCoverOfIsIso (𝟙 _))
+    (fun _ ↦ 𝒰) (fun _ _ _ ↦ hf _ _)
+
+theorem HasAffineProperty.affineProperty_diagonal (P) {Q} [HasAffineProperty P Q] :
+    Q.diagonal = AffineTargetMorphismProperty.of P.diagonal := by
+  ext X Y f _
+  exact of_affineProperty_diagonal P
+
+-- Doesn't build without the `IsAffine` instance but the linter complains
+@[nolint unusedHavesSuffices]
+instance HasAffineProperty.diagonal_affineProperty_isLocal
+    {Q : AffineTargetMorphismProperty} [Q.IsLocal] :
+    Q.diagonal.IsLocal where
+  respectsIso := inferInstance
+  to_basicOpen {X Y} _ f r hf :=
+    have : IsAffine (Y ∣_ᵤ Y.basicOpen r) := (isAffineOpen_top Y).basicOpen r
+    affineProperty_diagonal_of_diagonal_of_isPullback (targetAffineLocally Q)
+      (isPullback_morphismRestrict f (Y.basicOpen r)).flip
+      ((of_affineProperty_diagonal (targetAffineLocally Q)).mp hf)
+  of_basicOpenCover {X Y} _ f s hs hs' := by
+    refine (of_affineProperty_diagonal (targetAffineLocally Q)).mpr ?_
+    let 𝒰 := Y.openCoverOfSuprEqTop _ (((isAffineOpen_top Y).basicOpen_union_eq_self_iff _).mpr hs)
+    have (i) : IsAffine (𝒰.obj i) := (isAffineOpen_top Y).basicOpen i.1
+    refine diagonal_of_openCover_diagonal_affineProperty (targetAffineLocally Q) f
+      (Y.openCoverOfSuprEqTop _ (((isAffineOpen_top Y).basicOpen_union_eq_self_iff _).mpr hs)) ?_
+    intro i
+    exact (Q.diagonal.arrow_mk_iso_iff
+      (morphismRestrictEq _ (by simp) ≪≫ morphismRestrictOpensRange _ _)).mp (hs' i)
+#align algebraic_geometry.affine_target_morphism_property.is_local.diagonal AlgebraicGeometry.HasAffineProperty.diagonal_affineProperty_isLocal
+
+theorem HasAffineProperty.targetAffineLocally_diagonal_affineProperty
+    (P) {Q} [HasAffineProperty P Q] :
+    targetAffineLocally Q.diagonal = P.diagonal := by
```

Comment:
I think this should be `private` as it should only be used to construct `HasAffineProperty P.diagonal Q.diagonal`.